packages/@aws-cdk/core/lib/size.ts

```diff
@@ -0,0 +1,140 @@
+ * cannot be negative, or as an unresolved number token.
+ *
+ * When the amount is passed as a token, unit conversion is not possible.
+ */
+export class Size {
+  /**
+   * Create a Storage representing an amount kibibytes.
+   * 1 KiB = 1024 bytes
+   */
+  public static kibibytes(amount: number): Size {
+    return new Size(amount, StorageUnit.Kibibytes);
+  }
+
+  /**
+   * Create a Storage representing an amount mebibytes.
+   * 1 MiB = 1024 KiB
+   */
+  public static mebibytes(amount: number): Size {
+    return new Size(amount, StorageUnit.Mebibytes);
+  }
+
+  /**
+   * Create a Storage representing an amount mebibytes.
+   * 1 GiB = 1024 MiB
+   */
+  public static gibibytes(amount: number): Size {
+    return new Size(amount, StorageUnit.Gibibytes);
+  }
+
+  /**
+   * Create a Storage representing an amount tebibytes.
+   * 1 TiB = 1024 GiB
+   */
+  public static tebibytes(amount: number): Size {
+    return new Size(amount, StorageUnit.Tebibytes);
+  }
+
+  /**
+   * Create a Storage representing an amount pebibytes.
+   * 1 PiB = 1024 TiB
+   */
+  public static pebibyte(amount: number): Size {
+    return new Size(amount, StorageUnit.Pebibytes);
+  }
+
+  private readonly amount: number;
+  private readonly unit: StorageUnit;
+
+  private constructor(amount: number, unit: StorageUnit) {
+    if (!Token.isUnresolved(amount) && amount < 0) {
+      throw new Error(`Storage amounts cannot be negative. Received: ${amount}`);
+    }
+    this.amount = amount;
+    this.unit = unit;
+  }
+
+  /**
+   * Return this storage as a total number of kibibytes.
+   */
+  public toKibibytes(opts: StorageConversionOptions = {}): number {
+    return convert(this.amount, this.unit, StorageUnit.Kibibytes, opts);
+  }
+
+  /**
+   * Return this storage as a total number of mebibytes.
+   */
+  public toMebibytes(opts: StorageConversionOptions = {}): number {
+    return convert(this.amount, this.unit, StorageUnit.Mebibytes, opts);
+  }
+
+  /**
+   * Return this storage as a total number of gibibytes.
+   */
+  public toGibibytes(opts: StorageConversionOptions = {}): number {
+    return convert(this.amount, this.unit, StorageUnit.Gibibytes, opts);
+  }
+
+  /**
+   * Return this storage as a total number of tebibytes.
+   */
+  public toTebibytes(opts: StorageConversionOptions = {}): number {
+    return convert(this.amount, this.unit, StorageUnit.Tebibytes, opts);
+  }
+
+  /**
+   * Return this storage as a total number of pebibytes.
+   */
+  public toPebibytes(opts: StorageConversionOptions = {}): number {
+    return convert(this.amount, this.unit, StorageUnit.Pebibytes, opts);
+  }
+}
+
+/**
+ * Options for how to convert time to a different unit.
+ */
+export interface StorageConversionOptions {
+  /**
+   * If `true`, conversions into a larger storage units (e.g. `Kibibytes` to `Mebibytes`) will fail if the result is not
+   * an integer.
+   * @default true
+   */
+  readonly integral?: boolean;
```

Comment:
The semantics of `integral` seems a bit unintuitive. If I understand correctly, if `integral: false` the result will be rounded to an integer and if `integral: true` an error will be thrown for non-integral results. Correct?

Maybe it would make it clearer if it would be modeled as an enum `roundingBehavior`:

- `FLOOR` and
- `FAIL`

Or something like that.

Is there a use case for a floating point result (i.e. don't round)?